### PR TITLE
[chrome-ext] feat: Slack chat images work now

### PR DESCRIPTION
**Pull request by runpengl**
FFS images in chat finally work
Need to request files from Slack with auth token and make chrome extension source it with blob URL

slack-image-upload

### CODE REVIEW

File: chrome-extension/dist/inject_toolbar.js [hunk 1]
```diff
@@ -32,10 +32,11 @@ function injectToolbar() {
     // Some full-screen web-apps like google docs and slack don't respond well
     // to transform styles on the body, so we special-case their injection here.
     if (window.location.hostname === "docs.google.com") {
-        var docsChrome = document.getElementById("docs-chrome");
-        docsChrome.insertBefore(iframe, docsChrome.firstChild);
-
-        var chatContainer = document.querySelector(".docs-chat-pane-container");
+        const docsChrome = document.getElementById("docs-chrome");
+        if (docsChrome) {
```
Comment (vsiao) [inline chrome-extension/dist/inject_toolbar.js]: when does docs-chrome not exist?

File: chrome-extension/dist/inject_toolbar.js [hunk 2]
```diff
@@ -45,9 +46,11 @@ function injectToolbar() {
                 }
             });
         }
-    } else if (window.location.hostname === "superteamawesome.slack.com") {
-        var clientUi = document.getElementById("client-ui");
-        clientUi.insertBefore(iframe, clientUi.firstChild);
+    } else if (window.location.hostname === "app.slack.com") {
```
Comment (vsiao) [inline chrome-extension/dist/inject_toolbar.js]: oh, that's annoying. now the toolbar will show up even when you're using slack for work or something?

Comment (vsiao) [reply chrome-extension/dist/inject_toolbar.js]: can we only inject if the team id is T03A0NUTH?

File: chrome-extension/src/constants.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export default {
+    SLACK_TEAM_ID: 'T03A0NUTH'
```
Comment (vsiao) [inline chrome-extension/src/constants.js]: this seems like the sort of thing that should go in `config.js`

File: chrome-extension/src/chat/Message.js [hunk 4]
```diff
@@ -79,17 +81,8 @@ function renderTime(ts) {
...
-                ? <img
-                    className="Message-thumb"
-                    src={file.thumb_360_gif ? file.thumb_360_gif : file.thumb_360}
-                    width={Math.min(file.thumb_360_w, 300)}
-                    height={file.thumb_360_h * 300 / Math.max(300, file.thumb_360_w)}
-                />
-                : null}
-        </>;
+        const thumbUrl = file.thumb_360_gif ? file.thumb_360_gif : file.thumb_360;
+        return <MessageImage file={file} thumbUrl={thumbUrl} connectionInfo={props.connectionInfo} />
```
Comment (vsiao) [inline chrome-extension/src/chat/Message.js]: what about when the file isn't an image?

File: chrome-extension/src/chat/MessageImage.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        const { error, loading, imgBlob } = this.state;
+        const { thumbUrl, file } = this.props;
+
+        return (
+            <>
+                <div className="Message-fileName"><a href={file.url_private} target="_blank">{file.title}</a></div>
+                {error && <span className="Message-loadFail" style={{color: "orange"}}>Failed to load image</span>}
+                {loading && <span className="Message-loading" style={{color: "gray"}}>Loading...</span>}
+                {imgBlob && <img
+                        className="Message-thumb"
```
Comment (vsiao) [inline chrome-extension/src/chat/MessageImage.js]: can you make these `MessageImage-` prefixed? when we get .scss etc. set-up I want components and .scss files to have a 1:1 correspondence (and all classNames to be namespaced by their component name)

File: chrome-extension/dist/inject_toolbar.js (hunk 2 above)
Comment (runpengl) [reply chrome-extension/dist/inject_toolbar.js]: This function only gets executed if the team ID was also matched in another call

Comment (runpengl) [reply chrome-extension/dist/inject_toolbar.js]: Or what I mean is in `background.js` the toolbar is only rendered if the team ID matches

Comment (vsiao) [reply chrome-extension/dist/inject_toolbar.js]: yeah, makes sense

File: chrome-extension/src/background/background.js [hunk 6]
```diff
@@ -36,11 +36,16 @@ function handleChromeTabsUpdated(tabId, changeInfo, tab) {
...
                     // Try and inject a toolbar onto the page. If we have previously
                     // injected a toolbar this is a noop.
-                    chrome.tabs.executeScript(tabId, { file: "inject_toolbar.js" }, () => {
-                        console.log("chat widget state", ChatWidgetState.state());
-                        chrome.tabs.sendMessage(tabId, {
-                            msg: "injectChatWidget",
-                            chatWidgetState: ChatWidgetState.state()
+                    chrome.tabs.executeScript(tabId, {
+                        // hack method to pass config to `inject_toolbar`
+                        code: `var staConfig = ${JSON.stringify(config)}`
```
Comment (vsiao) [inline chrome-extension/src/background/background.js]: ehhhhh I don't know about this. as you say we don't even execute `inject_toolbar.js` if it's the wrong slack team. i'd rather keep it as you had it before.

File: chrome-extension/src/chat/Message.js [hunk 7]
```diff
@@ -79,17 +81,18 @@ function renderTime(ts) {
...
 function renderFiles(props) {
     return props.files.map(file => {
-        return <>
-            <div className="Message-fileName">{file.title}</div>
-            {file.thumb_160
-                ? <img
-                    className="Message-thumb"
-                    src={file.thumb_360_gif ? file.thumb_360_gif : file.thumb_360}
-                    width={Math.min(file.thumb_360_w, 300)}
-                    height={file.thumb_360_h * 300 / Math.max(300, file.thumb_360_w)}
```
Comment (vsiao) [inline chrome-extension/src/chat/Message.js]: I actually don't remember what was broken about this (and maybe a comment at the top of `MessageImage` would be a good reminder for future me too)

File: chrome-extension/src/chat/MessageImage.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            <>
+                <div className="MessageImage-fileName"><a href={file.url_private} target="_blank">{file.title}</a></div>
+                {error && <span className="MessageImage-loadFail" style={{color: "orange"}}>Failed to load image</span>}
+                {loading && <span className="MessageImage-loading" style={{color: "gray"}}>Loading...</span>}
+                {imgBlob && <img
+                        className="MessageImage-thumb"
+                        data-src={thumbUrl}
+                        src={imgBlob}
+                        onError={this.onError}
+                        onClick={() => { window.open(file.url_private); }}
```
Comment (vsiao) [inline chrome-extension/src/chat/MessageImage.js]: would rather we just wrap with `<a href` like you did with the title above

File: chrome-extension/src/chat/Message.js (hunk 7 above)
Comment (runpengl) [reply chrome-extension/src/chat/Message.js]: Yep, boils down to:
* Chrome extensions require external images to be sourced as Blob URLs
* Cross-site requests for Slack channel files require Oauth token to be included in Authorization header

**Pull request merged**